File: cqf-fhir-cql/src/main/java/org/opencds/cqf/fhir/cql/CqlEngineOptions.java
package org.opencds.cqf.fhir.cql;

import java.util.EnumSet;
import java.util.Set;
import java.util.StringJoiner;
import org.opencds.cqf.cql.engine.execution.CqlEngine;

// TODO: Eventually, the cql-engine needs to expose these itself.
public class CqlEngineOptions {
    private Set<CqlEngine.Options> options = EnumSet.noneOf(CqlEngine.Options.class);
    private boolean isDebugLoggingEnabled = false;
    private boolean shouldExpandValueSets = false;
    private Integer pageSize;
    private Integer maxCodesPerQuery;
    private Integer queryBatchThreshold;
    private boolean enableHedisCompatibilityMode = false;

    public Set<CqlEngine.Options> getOptions() {
        return this.options;
    }

    public void setOptions(Set<CqlEngine.Options> options) {
        this.options = options;
    }

    public boolean isDebugLoggingEnabled() {
        return this.isDebugLoggingEnabled;
    }

    public void setDebugLoggingEnabled(boolean isDebugLoggingEnabled) {
        this.isDebugLoggingEnabled = isDebugLoggingEnabled;
    }

    public boolean shouldExpandValueSets() {
        return this.shouldExpandValueSets;
    }

    public void setShouldExpandValueSets(boolean shouldExpandValueSets) {
        this.shouldExpandValueSets = shouldExpandValueSets;
    }

    public Integer getPageSize() {
        return this.pageSize;
    }

    public void setPageSize(Integer value) {
        this.pageSize = value;
    }

    public Integer getMaxCodesPerQuery() {
        return this.maxCodesPerQuery;
    }

    public void setMaxCodesPerQuery(Integer value) {
        this.maxCodesPerQuery = value;
    }

    public Integer getQueryBatchThreshold() {
        return this.queryBatchThreshold;
    }

    public void setQueryBatchThreshold(Integer value) {
        this.queryBatchThreshold = value;
    }

    public void setEnableHedisCompatibilityMode(boolean enableHedisCompatibilityMode) {
        this.enableHedisCompatibilityMode = enableHedisCompatibilityMode;
    }

    public boolean isEnableHedisCompatibilityMode() {
        return this.enableHedisCompatibilityMode;
    }

    public static CqlEngineOptions defaultOptions() {
        CqlEngineOptions result = new CqlEngineOptions();
        result.options.add(CqlEngine.Options.EnableExpressionCaching);
        return result;
    }

    @Override
    public String toString() {
        return new StringJoiner(", ", CqlEngineOptions.class.getSimpleName() + "[", "]")
                .add("options=" + options)
                .add("isDebugLoggingEnabled=" + isDebugLoggingEnabled)
                .add("shouldExpandValueSets=" + shouldExpandValueSets)
                .add("pageSize=" + pageSize)
                .add("maxCodesPerQuery=" + maxCodesPerQuery)
                .add("queryBatchThreshold=" + queryBatchThreshold)
                .add("enableHedisCompatibilityMode=" + enableHedisCompatibilityMode)
                .toString();
    }
}

File: cqf-fhir-cql/src/main/java/org/opencds/cqf/fhir/cql/CqlEngineOptions.kt
package org.opencds.cqf.fhir.cql

import java.util.StringJoiner
import org.opencds.cqf.cql.engine.debug.DebugMap
import org.opencds.cqf.cql.engine.execution.CqlEngine

// TODO: Eventually, the cql-engine needs to expose these itself.
class CqlEngineOptions {
    var options = mutableSetOf<CqlEngine.Options>()
    var isDebugLoggingEnabled: Boolean = false
    var debugMap: DebugMap? = null
    private var shouldExpandValueSets = false
    var pageSize: Int? = null
    var maxCodesPerQuery: Int? = null
    var queryBatchThreshold: Int? = null
    var isEnableHedisCompatibilityMode: Boolean = false

    var isProfilingEnabled: Boolean
        get() = this.options.contains(CqlEngine.Options.EnableProfiling)
        set(enabled) {
            if (enabled) {
                this.options.add(CqlEngine.Options.EnableProfiling)
            } else {
                this.options.remove(CqlEngine.Options.EnableProfiling)
            }
        }

    var isTracingEnabled: Boolean
        get() = this.options.contains(CqlEngine.Options.EnableTracing)
        set(enabled) {
            if (enabled) {
                this.options.add(CqlEngine.Options.EnableTracing)
            } else {
                this.options.remove(CqlEngine.Options.EnableTracing)
            }
        }

    var isCoverageEnabled: Boolean
        get() = this.options.contains(CqlEngine.Options.EnableCoverageCollection)
        set(enabled) {
            if (enabled) {
                this.options.add(CqlEngine.Options.EnableCoverageCollection)
            } else {
                this.options.remove(CqlEngine.Options.EnableCoverageCollection)
            }
        }

    var isDetailedTracingEnabled: Boolean
        get() = this.options.contains(CqlEngine.Options.EnableDetailedTracing)
        set(enabled) {
            if (enabled) {
                this.options.add(CqlEngine.Options.EnableDetailedTracing)
            } else {
                this.options.remove(CqlEngine.Options.EnableDetailedTracing)
            }
        }

    fun shouldExpandValueSets(): Boolean {
        return this.shouldExpandValueSets
    }

    fun setShouldExpandValueSets(shouldExpandValueSets: Boolean) {
        this.shouldExpandValueSets = shouldExpandValueSets
    }

    override fun toString(): String {
        return StringJoiner(", ", CqlEngineOptions::class.java.simpleName + "[", "]")
            .add("options=$options")
            .add("isDebugLoggingEnabled=$isDebugLoggingEnabled")
            .add("debugMap=" + (if (debugMap != null) "configured" else "null"))
            .add("shouldExpandValueSets=$shouldExpandValueSets")
            .add("pageSize=$pageSize")
            .add("maxCodesPerQuery=$maxCodesPerQuery")
            .add("queryBatchThreshold=$queryBatchThreshold")
            .add("enableHedisCompatibilityMode=" + this.isEnableHedisCompatibilityMode)
            .toString()
    }

    companion object {
        @JvmStatic
        fun defaultOptions(): CqlEngineOptions {
            val result = CqlEngineOptions()
            result.options.add(CqlEngine.Options.EnableExpressionCaching)
            return result
        }
    }
}
